include/ibamr/SnapshotCache.h

```diff
@@ -67,30 +73,51 @@ class SnapshotCache
      * are created than initially allocated. If you want to overwrite the snapshot, you should call updateSnapshot().
      * Data is not allocated in this class until this function is called.
      */
-    void setSnapshot(int u_idx, double time, SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy);
+    void
+    setSnapshot(int u_idx, double time, SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > current_hierarchy);
 
     /*!
      * Update a snapshot. The data from this patch index is copied into an array of indices controlled by this
      * object. This outputs an error if a snapshot with that time point within the specified tolerance can not be found.
      */
     void updateSnapshot(int u_idx,
                         double time,
-                        double tol,
-                        SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy);
+                        SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > current_hierarchy,
+                        double tol = 1.0e-8);
 
     /*!
      * Transfer the snapshot at the time value within a given tolerance to the patch index u_idx on the supplied patch
      * hierarchy.
      *
      * Note this can cause a significant amount of communication if the supplied patch hierarchy has a different
      * configuration of patches than the snapshot patch hierarchy.
+     *
+     * This function does not synchronize the hierarchy (i.e. no coarsening is performed). It copies the snapshot as it
+     * was stored.
```

Comment:
Does this mean that the caller needs to have the original hierarchy to copy back onto?

Reply:
I don't know about the specifics of copying data from one hierarchy to another, but whatever hierarchy you want to copy back onto has to be "valid" to copy to. I think that's just that the snapshotted hierarchy and the one to copy to have the same number of levels, their coarsest levels cover the same index space, and they have the same refinement ratios.

I'll figure out the details when I test.

Reply:
It seems like there are two likely use cases:

1. The caller wants the data on the original patch hierarchy layout; in this case, maybe the function should give access to a pointer to the internal clone of the patch hierarchy? (Most models will only have one `PatchHierarchy` object, and so recovering the hierarchy state at some previous time point seems annoying.)
2. The caller wants the data on their current patch hierarchy layout; in this case, maybe there could be a different function that uses something (`HierarchyGhostCellInterpolator`?) to transfer to the desired hierarchy layout.

Reply:
> 2\. The caller wants the data on their current patch hierarchy layout; in this case, maybe there could be a different function that uses something (`HierarchyGhostCellInterpolator`?) to transfer to the desired hierarchy layout.

This is what the function `getSnapshot` currently does: transfers from the snapshot hierarchy onto whatever patch hierarchy is given as the argument to the function. Although I realize there's an issue in the implementation that I'll fix soon.


> 1. The caller wants the data on the original patch hierarchy layout; in this case, maybe the function should give access to a pointer to the internal clone of the patch hierarchy? (Most models will only have one `PatchHierarchy` object, and so recovering the hierarchy state at some previous time point seems annoying.)

We can easily return the snapshotted patch hierarchy, perhaps in a function `getSnapshotAndHierarchy`.

Reply:
If you want to copy between hierarchies you might want to take a look at `IBTK::SecondaryHierarchy`. The original point of that class was to implement an alternative parallel partitioning but I think it implements what you need here too.